From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Mon, 26 May 2025 18:12:08 +0200
Subject: [PATCH] Expose low level steering API


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 74a2b3dcd303b25e9c26f3e4d4220b390b32f1d5..aa583dad42528bf523dd83f9f1e4f790aa0442cd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1553,7 +1553,12 @@ public abstract class Entity
         return colliders.isEmpty() ? maxDistance : -Shapes.collide(Direction.Axis.Y, aabb, colliders, -maxDistance);
     }
 
-    private Vec3 collide(final Vec3 movement) {
+    // KTP start - expose to static environment
+    public Vec3 collide(final Vec3 movement) {
+        return collideStatic(movement, this.level, this.getBoundingBox(), this);
+    }
+    public static Vec3 collideStatic(final Vec3 movement, final Level level, final AABB boundingBox, final @javax.annotation.Nullable Entity entity) {
+    // KTP end - expose to static environment
         // Paper start - optimise collisions
         final boolean xZero = movement.x == 0.0;
         final boolean yZero = movement.y == 0.0;
@@ -1562,7 +1567,7 @@ public abstract class Entity
             return movement;
         }
 
-        final AABB currentBox = this.getBoundingBox();
+        final AABB currentBox = boundingBox; // KTP - expose to static environment
 
         final List<VoxelShape> potentialCollisionsVoxel = new ArrayList<>();
         final List<AABB> potentialCollisionsBB = new ArrayList<>();
@@ -1579,11 +1584,11 @@ public abstract class Entity
 
         final List<AABB> entityAABBs = new ArrayList<>();
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getEntityHardCollisions(
-            this.level, (Entity)(Object)this, initialCollisionBox, entityAABBs, 0, null
+            level, (Entity)(Object)entity, initialCollisionBox, entityAABBs, 0, null // KTP - expose to static environment
         );
 
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, initialCollisionBox, potentialCollisionsVoxel, potentialCollisionsBB,
+            level, (Entity)(Object)entity, initialCollisionBox, potentialCollisionsVoxel, potentialCollisionsBB, // KTP - expose to static environment
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
         );
         potentialCollisionsBB.addAll(entityAABBs);
@@ -1597,7 +1602,7 @@ public abstract class Entity
 
         final double stepHeight;
 
-        if ((!collidedDownwards && !this.onGround) || (!collidedX && !collidedZ) || (stepHeight = (double)this.maxUpStep()) <= 0.0) {
+        if ((!collidedDownwards && !(entity == null ? false : entity.onGround)) || (!collidedX && !collidedZ) || (stepHeight = (double)(entity == null ? 0 : entity.maxUpStep())) <= 0.0) { // KTP - expose to static environment
             return collided;
         }
 
@@ -1611,7 +1616,7 @@ public abstract class Entity
         final List<AABB> stepAABBs = entityAABBs;
 
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, stepRetrievalBox, stepVoxels, stepAABBs,
+            level, (Entity)(Object)entity, stepRetrievalBox, stepVoxels, stepAABBs, // KTP - expose to static environment
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
         );
 
