From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Mon, 23 Jun 2025 14:51:56 +0200
Subject: [PATCH] Use alternative registry access if level null


diff --git a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
index b2b69fc2020990d11158e0b963d015c890e2518f..2ea2bbaab8a32ee755065b4194576532ff759fd9 100644
--- a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -153,7 +153,7 @@ public class JukeboxBlockEntity extends BlockEntity implements ContainerSingleIt
     public void setTheItem(ItemStack item) {
         this.item = item;
         boolean flag = !this.item.isEmpty();
-        Optional<Holder<JukeboxSong>> optional = JukeboxSong.fromStack(this.level.registryAccess(), this.item);
+        Optional<Holder<JukeboxSong>> optional = JukeboxSong.fromStack(this.level != null ? this.level.registryAccess() : org.bukkit.craftbukkit.CraftRegistry.getMinecraftRegistry(), this.item); // Paper - fallback to other RegistryAccess if no level
         this.notifyItemChangedInJukebox(flag);
         if (flag && optional.isPresent()) {
             this.jukeboxSongPlayer.play(this.level, optional.get());
