From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Sun, 22 Mar 2026 16:17:01 +0000
Subject: [PATCH] Increase command input limit default


diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
index 073e369fc21d1e098cb680638859a37b154895a4..2892043cf942e3b8121c80edda8cbff47024a0de 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandPacket.java
@@ -6,7 +6,7 @@ import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 
 public record ServerboundChatCommandPacket(String command) implements Packet<ServerGamePacketListener> {
-    public static final Integer MAX_CHAT_PACKET_INPUT_SIZE = Integer.getInteger("paper.maxChatCommandInputSize", 256); // Paper - limit chat command inputs
+    public static final Integer MAX_CHAT_PACKET_INPUT_SIZE = Integer.getInteger("paper.maxChatCommandInputSize", 256 * 10); // Paper - limit chat command inputs // KTP - increase limit
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatCommandPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatCommandPacket::write, ServerboundChatCommandPacket::new
     );
