From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Tue, 10 Jun 2025 16:19:29 +0200
Subject: [PATCH] Expose reduce debug info effect


diff --git a/src/main/java/org/bukkit/EntityEffect.java b/src/main/java/org/bukkit/EntityEffect.java
index 84d2a21a68c7f7244fb610103a08db5a4fdd55ce..df615681c639ef885d058fbec9a47b32e877dc6a 100644
--- a/src/main/java/org/bukkit/EntityEffect.java
+++ b/src/main/java/org/bukkit/EntityEffect.java
@@ -249,6 +249,8 @@ public enum EntityEffect {
      */
     GUARDIAN_TARGET(21, Guardian.class),
     // 22-28 player internal flags
+    PLAYER_REDUCE_DEBUG_INFO_TRUE(22, Player.class), // KTP - expose reduce debug info
+    PLAYER_REDUCE_DEBUG_INFO_FALSE(23, Player.class), // KTP - expose reduce debug info
     /**
      * Shield blocks attack.
      *
