From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Tue, 10 Jun 2025 16:19:29 +0200
Subject: [PATCH] Expose reduce debug info effect


diff --git a/src/main/java/org/bukkit/EntityEffect.java b/src/main/java/org/bukkit/EntityEffect.java
index cbf1d7a674f4010b3abb5c7fd3d2d8fcc60497ee..7dd47a862866e6be80da0adb4a7197ec9c8cddf9 100644
--- a/src/main/java/org/bukkit/EntityEffect.java
+++ b/src/main/java/org/bukkit/EntityEffect.java
@@ -245,6 +245,8 @@ public enum EntityEffect {
      */
     GUARDIAN_TARGET(21, Guardian.class),
     // 22-28 player internal flags
+    PLAYER_REDUCE_DEBUG_INFO_TRUE(22, Player.class), // KTP - expose reduce debug info
+    PLAYER_REDUCE_DEBUG_INFO_FALSE(23, Player.class), // KTP - expose reduce debug info
     /**
      * Shield blocks attack.
      *
