From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kieran Wallbanks <kieran.wallbanks@gmail.com>
Date: Mon, 23 Jun 2025 12:44:14 -0400
Subject: [PATCH] Adventure 4.23.0

Port of Paper PR #12690

diff --git a/src/main/java/org/bukkit/SoundCategory.java b/src/main/java/org/bukkit/SoundCategory.java
index 44fed7472ca2d9620b145695bdc65eef12ab9500..f4c6af9d6f39b89c96542391c5989f7be07be3a9 100644
--- a/src/main/java/org/bukkit/SoundCategory.java
+++ b/src/main/java/org/bukkit/SoundCategory.java
@@ -37,7 +37,7 @@ public enum SoundCategory implements Sound.Source.Provider {
             case PLAYERS -> Sound.Source.PLAYER;
             case AMBIENT -> Sound.Source.AMBIENT;
             case VOICE -> Sound.Source.VOICE;
-            case UI -> throw new UnsupportedOperationException("Waiting on adventure release for the UI sound source"); // todo adventure
+            case UI -> Sound.Source.UI;
         };
     }
 }
