From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Wed, 7 Jan 2026 21:56:01 +0000
Subject: [PATCH] Expose var int writing/reading


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 91eecf69975bd12eeba960c258bd096b32469307..8e42d9a4070fde982f4f3a83ac9b23ee9caf2552 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -282,5 +282,20 @@ public interface UnsafeValues {
             throw new java.lang.IllegalArgumentException("The passed entity could not be deserialized", exception);
         }
     }
+
+    /**
+     * Writes a var int to the passed data output.
+     *
+     * @param value  the int value to write.
+     * @param buffer the buffer.
+     */
+    void writeVarInt(int value, java.nio.ByteBuffer buffer);
+
+    /**
+     * Reads a var int to the passed data input.
+     *
+     * @param buffer the buffer to read from.
+     */
+    int readVarInt(java.nio.ByteBuffer buffer);
     // KTP stop
 }
