From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Wed, 7 Jan 2026 21:56:01 +0000
Subject: [PATCH] Expose var int writing/reading


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 954b02e09589b09491fccb945c841a51a1678e28..2a123a3bf9333d5837e16254543a41eb01426db2 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -433,5 +433,20 @@ public interface UnsafeValues {
             throw new java.lang.IllegalArgumentException("The passed entity could not be deserialized", exception);
         }
     }
+
+    /**
+     * Writes a var int to the passed data output.
+     *
+     * @param value  the int value to write.
+     * @param buffer the buffer.
+     */
+    void writeVarInt(int value, java.nio.ByteBuffer buffer);
+
+    /**
+     * Reads a var int to the passed data input.
+     *
+     * @param buffer the buffer to read from.
+     */
+    int readVarInt(java.nio.ByteBuffer buffer);
     // KTP stop
 }
