From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jatyn Stacy <jlee0964@gmail.com>
Date: Thu, 22 May 2025 17:20:03 -0400
Subject: [PATCH] Add faked player spectate target change method


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index d2d864f444bece36d7277fdf7250bd8779e62344..d97291b1d22acc9f3f950d14e1d9507933e2c62f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1128,6 +1128,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.getHandle().connection.send(new ClientboundRemoveMobEffectPacket(entity.getEntityId(), CraftPotionEffectType.bukkitToMinecraftHolder(type)));
     }
 
+    @Override
+    public void sendSpectateTargetChange(org.bukkit.entity.@Nullable Entity target) {
+        if (this.getHandle().connection == null) return;
+
+        this.getHandle().connection.send(
+            new net.minecraft.network.protocol.game.ClientboundSetCameraPacket(
+                target == null ? this.getHandle() :  ((CraftEntity) target).getHandle()
+            )
+        );
+    }
+
     @Override
     public WorldBorder getWorldBorder() {
         return this.clientWorldBorder;
