From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jatyn Stacy <jlee0964@gmail.com>
Date: Thu, 22 May 2025 17:20:03 -0400
Subject: [PATCH] Add faked player spectate target change method


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 7b165efeca72112d866861e038637bc3a2af9731..049d4322a2013d88fcae1a1a07694049dd51516a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1131,6 +1131,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.getHandle().connection.send(new ClientboundRemoveMobEffectPacket(entity.getEntityId(), CraftPotionEffectType.bukkitToMinecraftHolder(type)));
     }
 
+    @Override
+    public void sendSpectateTargetChange(org.bukkit.entity.@Nullable Entity target) {
+        if (this.getHandle().connection == null) return;
+
+        this.getHandle().connection.send(
+            new net.minecraft.network.protocol.game.ClientboundSetCameraPacket(
+                target == null ? this.getHandle() :  ((CraftEntity) target).getHandle()
+            )
+        );
+    }
+
     @Override
     public WorldBorder getWorldBorder() {
         return this.clientWorldBorder;
