From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Sun, 28 Sep 2025 14:59:37 +0200
Subject: [PATCH] Per-player glowing


diff --git a/src/main/java/io/papermc/paper/network/PaperSeenByTracker.java b/src/main/java/io/papermc/paper/network/PaperSeenByTracker.java
index ea08eceb04a8d956c11cb73ffbae6391642d3f1c..38192e6373acc6b882faca7c2e0476df71e81e6c 100644
--- a/src/main/java/io/papermc/paper/network/PaperSeenByTracker.java
+++ b/src/main/java/io/papermc/paper/network/PaperSeenByTracker.java
@@ -44,6 +44,32 @@ public class PaperSeenByTracker {
             this.connection = connection;
         }
 
+        // KTP start - glow override
+        public net.kyori.adventure.util.TriState glowOverride = net.kyori.adventure.util.TriState.NOT_SET;
+
+        public byte patchSharedFlagForGlowOverride(final byte value) {
+            return (byte) switch (glowOverride) {
+                case TRUE -> value | (1 << 6);
+                case FALSE -> value & ~(1 << 6);
+                case NOT_SET -> value;
+            };
+        }
+
+        public java.util.List<net.minecraft.network.syncher.SynchedEntityData.DataValue<?>> patch(
+            final java.util.List<net.minecraft.network.syncher.SynchedEntityData.DataValue<?>> input,
+            final int sharedFlagIndex
+        ) {
+            if (sharedFlagIndex == -1) return input;
+
+            final java.util.List<net.minecraft.network.syncher.SynchedEntityData.DataValue<?>> modifiedCopy = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(input);
+            modifiedCopy.set(sharedFlagIndex, net.minecraft.network.syncher.SynchedEntityData.DataValue.create(
+                net.minecraft.world.entity.Entity.DATA_SHARED_FLAGS_ID,
+                patchSharedFlagForGlowOverride((byte) modifiedCopy.get(sharedFlagIndex).value())
+            ));
+            return modifiedCopy;
+        }
+        // KTP end - glow override
+
     }
 
     // This implementation is based on the assumption that iteration is a far more common use case than addition and removal
@@ -70,6 +96,7 @@ public class PaperSeenByTracker {
         if (trackerData.effectiveIndex != TRACKER_DATA_NON_INITIALISED_INDEX) return;
         trackerData.effectiveIndex = TRACKER_DATA_NON_EFFECTIVE;
 
+        editGlowOverride(trackerData, connection.getPlayer().getBukkitEntity().glowOverride(this.trackedEntity.entity.getBukkitEntity())); // KTP - per player glow - required before editing visibility to ensure it is in the initial send
         editVisibility(connection, trackerData, connection.getPlayer().getBukkitEntity().canSee(this.trackedEntity.entity.getBukkitEntity()));
     }
 
@@ -174,4 +201,77 @@ public class PaperSeenByTracker {
     public boolean isEmpty() {
         return this.effective.isEmpty();
     }
+
+    // KTP start - glow override
+    public int hasAnyGlowOverrides = 0;
+
+    /**
+     * Edits the glow override of a server player on this tracker.
+     *
+     * @param connection   the server player instance.
+     * @param glowOverride the override tri state.
+     */
+    public void editGlowOverride(final ServerPlayerConnection connection, final net.kyori.adventure.util.TriState glowOverride) {
+        final TrackerData trackerData = this.theoretical.get(connection);
+        if (trackerData == null) return;
+
+        editGlowOverride(trackerData, glowOverride);
+    }
+
+    public void editGlowOverride(
+        final TrackerData trackerData,
+        final net.kyori.adventure.util.TriState glowOverride
+    ) {
+        if (trackerData.glowOverride == glowOverride) return;
+
+        // Two cases where we have to modify the any variable.
+        // Either, we currently are not an override and are editing to be one.
+        // Or we are currently one and are moving to not be one.
+        // THe above check for noop changes allows the following if statements without exhaustive checks for TRUE -> FALSE and FALSE -> TRUE cases.
+        if (trackerData.glowOverride == net.kyori.adventure.util.TriState.NOT_SET) {
+            ++this.hasAnyGlowOverrides;
+        } else if (glowOverride == net.kyori.adventure.util.TriState.NOT_SET) {
+            --this.hasAnyGlowOverrides;
+        }
+
+        // Update cached data
+        trackerData.glowOverride = glowOverride;
+    }
+
+    public void sendUpdatedGlowOverride(
+        final ServerPlayerConnection connection
+    ) {
+        this.sendUpdatedGlowOverride(connection, this.theoretical.get(connection));
+    }
+
+    public void sendUpdatedGlowOverride(
+        final ServerPlayerConnection connection,
+        final @org.jspecify.annotations.Nullable TrackerData trackerData
+    ) {
+        if (trackerData == null || trackerData.effectiveIndex < 0) return;
+
+        // Actually send updated faked data to the player
+        final byte value = trackerData.patchSharedFlagForGlowOverride(
+            this.trackedEntity.entity.getEntityData().getItem(net.minecraft.world.entity.Entity.DATA_SHARED_FLAGS_ID).getValue()
+        );
+
+        connection.send(new net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket(
+            this.trackedEntity.entity.getId(),
+            java.util.List.of(net.minecraft.network.syncher.SynchedEntityData.DataValue.create(net.minecraft.world.entity.Entity.DATA_SHARED_FLAGS_ID, value))
+        ));
+    }
+
+    public int findSharedDataValueIndexIn(final java.util.List<net.minecraft.network.syncher.SynchedEntityData.DataValue<?>> list) {
+        for (int i = 0; i < list.size(); i++) {
+            final net.minecraft.network.syncher.SynchedEntityData.DataValue<?> dataValue = list.get(i);
+            if (dataValue.id() >= net.minecraft.world.entity.Entity.DATA_SHARED_FLAGS_ID.id()) {
+                if (dataValue.id() == net.minecraft.world.entity.Entity.DATA_SHARED_FLAGS_ID.id()) {
+                    return i;
+                }
+                break; // We can break here, if the data value id is higher than the shared flag id, the list does not contain the shared flags.
+            }
+        }
+        return -1;
+    }
+    // KTP end - glow override
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index aed888eb7f287fdb7a5f05962129c23f8b99d830..84e606945d20b82abf61165f125dabb30fcbf253 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3387,4 +3387,40 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             entity.remove();
         }
     }
+
+    // KTP start - glow override
+    private final it.unimi.dsi.fastutil.objects.Object2BooleanOpenHashMap<UUID> glowOverrides = new it.unimi.dsi.fastutil.objects.Object2BooleanOpenHashMap<>();
+    @Override
+    public void glowOverride(
+        final org.bukkit.entity.@org.jspecify.annotations.NonNull Entity entity,
+        final net.kyori.adventure.util.@org.jspecify.annotations.NonNull TriState glowOverride
+    ) {
+        if (glowOverride == TriState.NOT_SET) {
+            glowOverrides.removeBoolean(entity.getUniqueId());
+        } else {
+            glowOverrides.put(entity.getUniqueId(), Boolean.TRUE.equals(glowOverride.toBoolean()));
+        }
+
+        final Entity handleRaw = ((CraftEntity) entity).getHandleRaw();
+        final ChunkMap.TrackedEntity trackedEntity = handleRaw.moonrise$getTrackedEntity();
+        if (trackedEntity != null) {
+            trackedEntity.seenBy.editGlowOverride(this.getHandle().connection, glowOverride);
+            trackedEntity.seenBy.sendUpdatedGlowOverride(this.getHandle().connection);
+        }
+    }
+
+    /**
+     * Yields the glow override for the passed entity.
+     *
+     * @param entity the entity instance to check for.
+     *
+     * @return the tri state, {@code NOT_SET} if no override is configured.
+     */
+    @Override
+    public net.kyori.adventure.util.TriState glowOverride(final org.bukkit.entity.Entity entity) {
+        return this.glowOverrides.containsKey(entity.getUniqueId())
+                ? net.kyori.adventure.util.TriState.byBoolean(this.glowOverrides.getBoolean(entity.getUniqueId()))
+                : net.kyori.adventure.util.TriState.NOT_SET;
+    }
+    // KTP end - glow override
 }
