From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Wed, 7 Jan 2026 21:56:11 +0000
Subject: [PATCH] Expose var int writing/reading


diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 959d91ca0ee70f145bfa87b9717457ef01ebad1b..f610dce7206e76619f2b068b0225429a48285526 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -658,6 +658,18 @@ public final class CraftMagicNumbers implements UnsafeValues {
             CraftItemStack.asNMSCopy(itemStack)
         ).getOrThrow().toString();
     }
+
+    @Override
+    public void writeVarInt(final int value, final java.nio.ByteBuffer buffer) {
+        final io.netty.buffer.ByteBuf nettyWrapper = io.netty.buffer.Unpooled.wrappedBuffer(buffer);
+        nettyWrapper.clear();
+        net.minecraft.network.VarInt.write(nettyWrapper, value);
+    }
+
+    @Override
+    public int readVarInt(final java.nio.ByteBuffer buffer) {
+        return net.minecraft.network.VarInt.read(io.netty.buffer.Unpooled.wrappedBuffer(buffer));
+    }
     // KTP stop
 
     @Override
