From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Wed, 7 Jan 2026 21:56:11 +0000
Subject: [PATCH] Expose var int writing/reading


diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 12271d2f885cf857d0a59d43a147ab9dfa39aaa7..8ec7086013a856b7b40f974bf138bc43f55b4d37 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -602,6 +602,18 @@ public final class CraftMagicNumbers implements UnsafeValues {
             CraftItemStack.asNMSCopy(itemStack)
         ).getOrThrow().toString();
     }
+
+    @Override
+    public void writeVarInt(final int value, final java.nio.ByteBuffer buffer) {
+        final io.netty.buffer.ByteBuf nettyWrapper = io.netty.buffer.Unpooled.wrappedBuffer(buffer);
+        nettyWrapper.clear();
+        net.minecraft.network.VarInt.write(nettyWrapper, value);
+    }
+
+    @Override
+    public int readVarInt(final java.nio.ByteBuffer buffer) {
+        return net.minecraft.network.VarInt.read(io.netty.buffer.Unpooled.wrappedBuffer(buffer));
+    }
     // KTP stop
 
     @Override
