From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <lynxplay101@gmail.com>
Date: Wed, 21 Jul 2021 02:32:45 +0200
Subject: [PATCH] KPT pom changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 682935762008602ca214f68147766792cbedeea9..98f58f7593f9f2e8e534ffb539b81552f24dbd43 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -27,11 +27,14 @@ repositories {
         }
     }
     // Paper end
+    // KTP start
+    maven("https://papermc.io/repo/repository/maven-public/")
+    // KTP end
 }
 
 dependencies {
-    implementation(project(":Paper-API"))
-    implementation(project(":Paper-MojangAPI"))
+    implementation(project(":KTP-API"))
+    implementation("io.papermc.paper:paper-mojangapi:1.17.1-R0.1-SNAPSHOT")
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.12.1")
     implementation("net.minecrell:terminalconsoleappender:1.2.0")
@@ -81,7 +84,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Paper-$implementationVersion",
+            "Implementation-Version" to "git-ktp-$implementationVersion",
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
